nimare/meta/cbmr.py

```diff
@@ -0,0 +1,783 @@
+        Contrast matrix for homogeneity test or group comparison on estimated spatial
+        intensity function.
+        For boolean inputs, no statistical inference will be conducted for spatial intensity
+        if `t_con_group` is False, and spatial homogeneity test for groupwise intensity
+        function will be conducted if `t_con_group` is True.
+        For list inputs, generialized linear hypothesis (GLH) testing will be conducted for
+        each element independently. We also allow any element of `t_con_group` in list type,
+        which represents GLH is conducted for all contrasts in this element simultaneously.
+        Default is homogeneity test on group-wise estimated intensity function.
+    t_con_moderators : :obj:`~bool` or obj:`~list` or obj:`~None`, optional
+        Contrast matrix for testing the existence of one or more study-level moderator effects.
+        For boolean inputs, no statistical inference will be conducted for study-level moderators
+        if `t_con_moderators` is False, and statistical inference on the effect of each study-level
+        moderators will be conducted if `t_con_group` is True.
+        For list inputs, generialized linear hypothesis (GLH) testing will be conducted for
+        each element independently. We also allow any element of `t_con_moderators` in list type,
+        which represents GLH is conducted for all contrasts in this element simultaneously.
+        Default is statistical inference on the effect of each study-level moderators
+    device: :obj:`string`, optional
+        Device type ('cpu' or 'cuda') represents the device on which operations will be allocated.
+        Default is 'cpu'.
+    """
+    
+    def __init__(self, CBMRResults, device="cpu"): 
+        self.device = device
+        self.CBMRResults = CBMRResults
+        self.groups = self.CBMRResults.estimator.groups
+        self.n_groups = len(self.groups)
+        
+        # visialize group/moderator names and their indices in contrast array
+        self.group_reference_dict, self.moderator_reference_dict = dict(), dict()
+        LGR.info("Group Reference in contrast array")
+        for i in range(self.n_groups):
+            self.group_reference_dict[self.groups[i]] = i
+            LGR.info(f"{self.groups[i]} = index_{i}")
+        if self.CBMRResults.estimator.moderators:
+            n_moderators = len(self.CBMRResults.estimator.moderators)
+            LGR.info("Moderator Reference in contrast array")
+            for j in range(n_moderators):
+                self.moderator_reference_dict[self.CBMRResults.estimator.moderators[j]] = j
+                LGR.info(f"{self.CBMRResults.estimator.moderators[j]} = index_{j}")
+            
+        # device check
+        if self.device == "cuda" and not torch.cuda.is_available():
+            LGR.debug("cuda not found, use device 'cpu'")
+            self.device = "cpu"
+
+    def create_contrast(self, contrast_name, type="group"):
```

Comment:
I think create_contrast should only create one contrast, if people want multiple contrasts, they can call the function multiple times.

instead of `homo_test_`, what is ambiguous about just including the group name? (e.g., "Schizophrenia_Yes" versus "homo_test_Schizophrenia_Yes"?)

instead of VS, did we consider using "+" and "-", and doing a split on those symbols and interpreting those symbols to decide which group gets (1, -1)?

Reply:
basing these comments around how nilearn plots contrasts: https://nilearn.github.io/dev/modules/generated/nilearn.plotting.plot_contrast_matrix.html